### ISSUE TITLE
Too hard to define a custom command wrapping a RPC

### ISSUE THREAD

**Issue by mitar** [cypress-io/cypress#2443]
### Current behavior:

Given existing documentation and reading the source code I was trying today the whole day to define a custom command wrapping a simple node-style callback RPC. In my view this is much harder than it should be if you want a perfect integration and native (like official commands) feel to this command.

First, it seems that [very important Cypress.log](https://docs.cypress.io/api/cypress-api/cypress-log.html) is barely documented and one has to go through the source code to really understand it, where also a very limited comments do not help.

Then there are the following issues:
* It seems it is not possible to get same treatment in log as XHR request, to get loading icon to show while the call is in progress. I have tried to setup log and then use `log.end()` call to signal when is start and end if the request, but it does not seem to have the effect.
* It is not really possible to hide internal commands. I tried to use `{log: false}` everywhere, but not all commands support that (like `its` I used), or if they do, internal commands still leak, for example, if my RPC call takes too long, then the error message about timeout comes from `then` instead of my custom command.
* While `log.error()` marks the entry in the log red, it does not really make the whole process fail. 
* It is unclear how one can return a subject in an async manner. An example in documentation would be great. In my case, I would like to return of the RPC as the subject from the command. I found `verifyUpcomingAssertions` but I am not sure how to use it, there was no documentation for it I could find.
* It is unclear how one can define a custom command which is partially retriable (like fetching the `RPC` object from `window`) but not fully (like I want RPC to be done only once).

This so to where I got:

```js
Cypress.Commands.add('call', (methodName, ...args) => {
  const options = {};

  options._log = Cypress.log({
    name: 'call',
    message: `${methodName}(${Cypress.utils.stringify(args)})`,
    consoleProps() {
      return {
        methodName,
        arguments: args,
      };
    },
  });

  return cy.window({log: false}).its('RPC').then({log: false}, (RPC) => {
    try {
      return new Promise((resolve, reject) => {
        RPC.call(methodName, ...args, (error, result) => {
          options._log.set({
            consoleProps() {
              return {
                methodName,
                arguments: args,
                error,
                result,
              };
            },
          });

          if (error) {
            options._log.error(error);
          }
          else {
            options._log.end();
            cy.verifyUpcomingAssertions(result, options);
          }
          resolve();
        });
      });
    }
    catch (error) {
      options._log.set({
        consoleProps() {
          return {
            methodName,
            arguments: args,
            error,
          };
        },
      }).error(error);
    }
  });
});
```

### Versions

Cypress 3.1.0.

**Comment by jennifer-shehane** [comment]
Thank you for the thoughtful write-up. I agree with all of your points and think most of them would be solved by [better documentation of the Cypress.log command](https://github.com/cypress-io/cypress-documentation/issues/110). 

Looped in some of the team to hopefully get some of your current questions answered.

**Comment by egucciar** [comment]
@mitar I found after significant digging my mistake was not returning legitimate `Promise` from the command. I use `cy.state('window')` now to retrieve window rather than rely on `cy.window` which will not currently invoke a spinner! Thx for raising this issue and inspiring me to dig deeper.

**Comment by mitar** [comment]
@jennifer-shehane Any official update on this?

**Comment by mitar** [comment]
My current attempt:

```js
Cypress.Commands.add('call', (methodName, ...args) => {
  const log = Cypress.log({
    name: 'call',
    message: `${methodName}(${Cypress.utils.stringify(args)})`,
    consoleProps() {
      return {
        methodName,
        arguments: args,
      };
    },
  });

  return new Promise((resolve, reject) => {
    const RPC = cy.state('window').RPC;

    RPC.call(methodName, ...args, (error, result) => {
      log.set({
        consoleProps() {
          return {
            methodName,
            arguments: args,
            error,
            result,
          };
        },
      });

      if (error) {
        log.error(error);
        reject(error);
      }
      else {
        resolve(result);
      }
    });
  });
});
```

It is not yet perfect. I get now double `call` error displayed, but it is better than nothing.

**Comment by egucciar** [comment]
@mitar take a look at my code example (specifically the error handling) to remove duplicate error

**Comment by mitar** [comment]
Thanks. It really works better:

```js
Cypress.Commands.add('call', (methodName, ...args) => {
  const log = Cypress.log({
    name: 'call',
    message: `${methodName}(${Cypress.utils.stringify(args)})`,
    consoleProps() {
      return {
        methodName,
        arguments: args,
      };
    },
  });

  return new Promise((resolve, reject) => {
    const RPC = cy.state('window').RPC;

    RPC.call(methodName, ...args, (error, result) => {
      log.set({
        consoleProps() {
          return {
            methodName,
            arguments: args,
            error,
            result,
          };
        },
      });

      if (error) {
        reject(error);
      }
      else {
        resolve(result);
      }
    });
  }).catch((error) => {
    Cypress.utils.throwErr(error, {
      onFail: log,
    });
  });
});
```

**Comment by mitar** [comment]
Ah, `cy.state('window')` does not work in `beforeEach`.

**Comment by egucciar** [comment]
That's correct, @mitar, it's probably a good idea to structure making these calls after visit.

**Comment by egucciar** [comment]
@mitar one more thing i found today was how to do "retryability". Let me know if you need these features, i have some insights.

**Comment by mitar** [comment]
For me is that I would like to prevent it. :-) I want the RPC call to be done only once.

**Comment by egucciar** [comment]
Well then there's no need to "prevent" it perse. Right? "Retryability" doesn't come for free. :)

**Comment by egucciar** [comment]
@mitar well it depends if you're chaining your command from let's say a `.get` cypress will call `verifyUpcomingAsserions` so things will re run until successful. I'd be suspicious if your code was auto retrying if you haven't chained it from something that auto retries.

**Comment by jennifer-shehane** [comment]
I am moving this issue to our [docs](https://github.com/cypress-io/cypress-documentation).
